---
title: "Informatics Challenge! 2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo=TRUE, 
  dev="svg", 
  fig.width=4, 
  fig.height=4,
  fig.asp=NULL
)

library(tidyverse)
library(ggplot2)
library(ggpubr)
library(showtext)
library(cowplot)
library(ggsci)

font_add("Arial Narrow", "ARIALN.TTF", bold = "ARIALNB.TTF")
showtext_auto()

theme_set(theme_pubr())
```

# How the informatics challenge will work

Welcome to the 2022 annual informatics challenge!

Below are a series of questions of varying difficulty. Each question will be
worth 1 - 5 points (denoted in the section header) before the questions. Some of
these questions are language specific, while others can be completed however
you'd like. Either way, **please show your work/code**. I've tried to specify in
each header if the questions are language specific, but feel free to ask for
clarification if needed.

I'm writing the questions to be either a task or specific question, such that
it is unambiguous if the answer is right. If you do the task or answer the
question correctly you get full points. No partial credit.

I am asking participants to specify their experience level so I can split folks
into two categories. I'll be awarding two trophies, one for each category.

-  Beginners I'm generally defining as those that are just getting started and have
been doing informatics regularly for a year or less.
-  You should likely submit as Advanced if you have been doing this for a while
and if all the beginner questions seem really easy and the harder questions
don't seem too bad.

There are several questions, so I'm giving two weeks to complete everything. You
don't have to complete all the questions, I'll be defining winners of the two
categories based on the highest total points. If you're the only one who enters,
you can win by answering just one question!

You can put your answers in-line in the Rmd file or you can provide your answers
another way if you prefer.

**The due date will be April 28th at 11:59 PM**

Feel free to ask for clarification on anything!

------------------------------------------------------------------------

## Are you submitting in the beginner or advanced category?

Advanced

------------------------------------------------------------------------

# Beginner questions - 1 point per question

## --R-specific-- Data and variable types

1.  Make a numeric vector at least three elements long

```{r}
num_vector <- c(1, 10, 100)
num_vector
```

2.  Make a named list object where the values are 10, 15, 20 and 25 and the
names of the elements are "Billy", "Eli", "Ali" and "Jordan"

```{r}
list_obj <- list(Billy = 10, Eli = 15, Ali = 20, Jordan = 25)
list_obj
```

3.  Make a data frame where one column is random foods and the second column is
made-up calorie values

```{r}
df <- tribble(
  ~foods, ~calories,
  "salad", 750,
  "nachos", 15,
  "grapes", 499,
  "pickles", 42,
  "butterflies", 7851
)
df
```

4.  Add a new column of some sort to the data frame you made in question 3 and
assign the output a new variable

```{r}
df$two_servings <- df$calories * 2
df
```

5.  Remove the column you just added and assign this to a new variable

```{r}
two_servings <- df$two_servings
two_servings

df <- df[,1:2]
df
```

## Make basic plots

You can access a test dataset in R using the following code:

```{r}
mtcars <- mtcars
```

If you want to use a language other than R, the data can be downloaded here: https://nationwidechildrens-my.sharepoint.com/:t:/g/personal/matthew_cannon_nationwidechildrens_org/EbYpZRiuizFOjd9nz4vKn38Br8z2hEKnRijltqbzsUgnow?e=pPnNxW

6.  Using the mtcars data, make a scatterplot of displacement (disp) vs mpg

```{r}
ggplot(mtcars, aes(disp, mpg)) +
  geom_point() +
  labs(
    title = "Displacement vs Efficiency",
    subtitle = "Bigger engines suck more gas",
    x = "Engine Displacement (ml)",
    y = "Fuel Efficiency (mpg)"
  )
```

7.  Using the mtcars data again, make a boxplot of hp separated by cyl

```{r, fig.width=3.5}
ggplot(mtcars, aes(as.factor(cyl), hp, fill = as.factor(cyl))) +
  geom_boxplot() +
  labs(
    title = "Power by Configuration",
    subtitle = "Want a hot rod? Get a V8",
    x = "Configuration (no. of cylinders)",
    y = "Power Output (hp)"
  )+
  scale_fill_jco() +
  theme(legend.position = "none")
```

8.  Using the mtcars data again, make a barplot of displacement (disp) for each
engine shape (column "vs")

```{r, fig.width=3}
mtcars_summary <- mtcars %>%
  group_by(vs) %>%
  summarise(
    sd = sd(disp),
    disp = mean(disp)
  )
mtcars_summary$vs <- c("V Shape", "Other Shape")

ggplot(mtcars_summary, aes(as.factor(vs), disp, fill = vs)) +
  geom_bar(stat = "identity", color = "gray20") +
  geom_errorbar(aes(
    ymin = disp - sd,
    ymax = disp + sd),
    width = 0.2,
    color = "gray20"
  ) +
  labs(
    title = "Piston Arrangments",
    subtitle = "Big cars like V-shaped engines",
    x = "Cylinder Arrangement",
    y = "Engine Displacement (ml)"
  ) +
  scale_fill_jco() +
  theme(legend.position = "none")
```

## Do a t-test

9.  What is the *p*-value for a *t*-test (using the mtcars data), that tests if
displacement (disp) varies between engine shapes (column "vs")?

```{r}
t.test(mtcars$disp ~ mtcars$vs)
```

## Basic unix commands

10.  Which command is used to create a new directory?

```{bash}
mkdir newDirectory
```

11.  Which command shows the size of a file?

```{bash}
stat informatics_challenge_2022.Rmd
ls -l
```

12.  Which command lets you delete a file?

```{bash}
touch file.txt
rm file.txt
```

13.  Which command copies a file?

```{bash}
touch file_orig.txt
cp file_orig.txt file_copy.txt
```

14.  Which command moves a file from one directory to another?

```{bash}
mv file_*.txt newDirectory
rm -r newDirectory
```

## Basic stats knowledge

15.  Does a *t*-test assume that the measurements follow a normal distribution?

Always

16.  Briefly (1-2 sentences), what is the difference between parametric and
non-parametric tests?

Parametric tests depend on comparisons of two different distributions.
Non-parametric tests operate indepented of any defined distribution, using
instead properties like rank order or other summary statistics.

17.  Name one method to perform multiple testing correction

Bonferroni (corrected p), Benjamini-Hochberg (FDR, q-value)

18.  For a set of statistical tests, using a FDR cutoff of 0.1 will, in theory,
result in what percentage of false positives in your "significant" tests?

Trick question... depends on how many comparisons you have.
Each positive hit will have a 0.1 probability of being a false positive.

## Basic file/data formats

19.  What file/data format is this?

```{verbatim, max.height='100px'}
>Sequence_1
AGTACATGGGGAGCTGGCGCAATGACGACCTAATTGGCGCACAGTACTAGGCATCGTCATCCAATGCGACGAGTCCTAC
ACTATCTTGGATATGATATGGCGCACTACACATGCTAGCCGCTGGGGAGATTAGCTCGAGTTGCCCCTTTGCCCGATCC
CGGAAGATACGCTCTAAGCTCGGCAATCGCTCTTGCCGTGCGAGATGCTAGCAAAAAGGTGTACTTCTCAGCGGAGCAG
AAAGATCATGTTTATTGGAAGCATCAACCTGCGCCGTCTTGTTAACTTGTCATATCGCGCACGTAGT
>Sequence_2
GCACTACACATGCTAGCCGCTGGGGAGATTAGCTCGAGTTGCCCCTTTGCCCGATCCCGGAAGATACGCTCTAAGCTCG
GCAATCGCTCTTGCCGTGCGAGATGCTAGCAA
```

.fastq, .fq
#MVC###########################################################################
# This is actually a fasta, in this format there aren't quality scores
# https://support.illumina.com/bulletins/2016/04/fastq-files-explained.html

20.  What file/data format is this?

```{verbatim, max.height='100px'}
@A00498:77:HGHFVDRXX:1:2101:3775:1016 1:N:0:NACGGCTC
ANTTAGGCAACCGTGCTTATGGAAATGCA
+
F#FFFFFFFFFFFFFFFFFFFFFFF:FF,
@A00498:77:HGHFVDRXX:1:2101:3830:1016 1:N:0:NATCGCTC
TNGCTGCAGAATAACCCGGGCTCAGACTC
+
F#FFFFFFFF,FFFFFFFFFFFFFFFFF:
```

.sam (or extracted .bam)
#MVC######################################################################
# This is fastq format. The sam format has one line per sequence with
# additional fields to hold the mapping location/quality and other data
# https://samtools.github.io/hts-specs/SAMv1.pdf


21.  What file/data format is this?

```{}
chr1  213941196  213942363
chr1  213942363  213943530
chr1  213943530  213944697
chr2  158364697  158365864
chr2  158365864  158367031
chr3  127477031  127478198
```

It's a chromosomal location file identifying locations within a particular reference
(some kind of an alignment).
#MVC#########################################################################
# Sortof, this is the bed format, defined as chromosome, start and stop. It
# can also have additional optional fields to hold different types of data

22.  What are the two main differences between a bam and sam file?

.bam is stored as a compressed binary, where .sam is stored as human-readable text

------------------------------------------------------------------------

# Intermediate - 2 points per question

## Make fancy plots

23.  Using the mtcars data, make a plot of mpg vs disp, coloring the points by
cyl. Include a title. Be sure the colors used are color-blind friendly.

```{r}
p <- ggplot(mtcars, aes(disp, mpg, color = as.factor(cyl))) +
  geom_point() +
  labs(
    title = "Efficiency by Volume",
    subtitle = "Bigger cars suck more gas",
    x = "Engine Displacment (ml)",
    y = "Fuel Efficiency (mpg)"
  ) +
  theme(legend.position = "bottom") +
  scale_color_jco(name = "Cylinders")
print(p)
```

24.  From the previous plot, include a legend for point colors and make the
title of the legend "Cylinder".

Oops! I'm ahead of you... already done.

25.  Add a linear regression line to the plot along with text in the plot 
providing the Pearson's correlation and the *p*-value

```{r}
t <- cor.test(mtcars$disp, mtcars$mpg, method = "pearson")
p + geom_smooth(
  aes(disp, mpg, group = 1),
  method = "lm",
  color = "black"
  ) +
  annotate("text", x = 375, y = 31,
    label = paste("p =", format(t$p.value, 0, 4), "(Pearson)")
  )
```

## Align reads to a reference

26.  On the Franklin cluster, in the folder /gpfs0/scratch/mvc002/ are two files:

  -   infoChallenge_R1.fastq.gz
  -   infoChallenge_R2.fastq.gz
  
  Align these reads to the hg38 reference using your aligner of choice and share
  your code.
  Several pre-made references are available here:
  `/reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/`.

27.  Share a screenshot of some of the aligned reads in IGV

28.  Is this dataset whole genome sequencing, RNAseq, ChIPseq or amplicon seq?

## --R-specific-- Demonstrate the use of pipes and tidyverse functions

29.  In R, starting with the mtcars dataset use `filter()` to remove any row
where cyl is "8" and then use `group_by()` to group the data by gear and then
use `summarize()` to calculate the average of the mpg column per group. This
should be in one continuous pipeline without intermediate variables or nested
functions. Use either pipe (`%>%` or `|>`).

```{r}
mpg_per_gear <- mtcars %>%
  filter(cyl !=8) %>%
  group_by(gear) %>%
  summarize(avg_mpg = mean(mpg))
mpg_per_gear
```

## Write a bash loop

30.  Write a "for loop" in bash to loop over the numbers 1-10 and print them to
the console

```{bash}
for i in {1..10}
  do
    echo "Number $i"
  done
```

## Git and GitHub

31.  For your submission, put your results up into a public GitHub repository
and provide the link

https://github.com/ryrobert/2022NCHCodingChallenge

32.  Make sure your repo includes at least two branches and two commits

Done

------------------------------------------------------------------------

# Hard - 3 points per question

## Harder plotting

33.  Add histograms to the top and right side of the mtcars mpg vs disp
scatterplot so that it looks like Figure 1. Note that the axes of each plot
should align.
#MVC########################################################################
# So *technically* you did density plots and not histograms. I'm not a big
# fan of density plots in general since they tend to over-smooth the data

```{r}
dx <- axis_canvas(p, axis = "x") +
  geom_density(data = mtcars, 
    aes(x = disp, fill = as.factor(cyl)), color = "gray50", alpha = 0.75) +
  scale_fill_jco()

dy <- axis_canvas(p, axis = "y", coord_flip = T) +
  geom_density(data = mtcars,
    aes(x = mpg, fill = as.factor(cyl)), color = "gray50", alpha = 0.6) +
  coord_flip() +
  scale_fill_jco()

p1 <- insert_xaxis_grob(p, dx, position = "top")
p1 <- insert_yaxis_grob(p1, dy, position = "right")
ggdraw(p1)
```

## Get sequences from SRA

34.  Using the command line, download the SRR10197556 dataset from the NCBI SRA
repository

```{bash, eval=F}
ml load SRA-Toolkit/2.10.8
prefetch SRR10197556
fasterq-dump SRR10197556
```

## Bash variable expansion

35.  Create a variable in bash containing the text "path/to/target/file.txt" and
then share code to trim off the path ("path/to/target/")

36.  Create a variable in bash containing the text "path/to/target/file.txt" and
then share code to trim off the file extension (".txt")

```{bash}
long="path/to/target/file.txt"
fname="${long##*/}"
echo "Answer to 35: ${fname}"
noext=${fname%.*}
echo "Answer to 36: ${noext}"
```

## Grep

37.  Use the Unix grep function to match the pattern "GATACA" in `/gpfs0/scratch/mvc002/infoChallenge_1.fastq.gz` on the Franklin cluster. How
many lines had a match?

```{bash}
zgrep -c "GATACA" /gpfs0/scratch/mvc002/infoChallenge_1.fastq.gz
```

## Write a program

38.  Share code written in a language other than R or bash that reads in a fastq
file and trims 10 characters from the 3' end of the read and quality score
fields. You can use the `/gpfs0/scratch/mvc002/infoChallenge_1.fastq.gz` file to
test your code. 

------------------------------------------------------------------------

# Too hard - 4 points per question

## Finding a sequence in a fasta file?

39.  How many times does sequence "GATTACA" appear in this fasta file? `/reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/WholeGenomeFasta/genome.fa`

```{bash}
grep -o "GATTACA" /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Sequence/WholeGenomeFasta/genome.fa | grep -c ""
```
#MVC###########################################################################
# Close, but this counts the number of lines that have at least one match, but
# doesn't count multiple matches per line. This approach will also miss any matches
# that occur where the sequence is wrapped to the next line and so the pattern
# is split across multiple lines of a single fasta entry

## Using awk

40.  Use this gtf file (`/reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Annotation/Genes/genes.gtf`)
and the Unix `awk` command to keep only the lines where first column is "chr11"
and the fourth column is greater than 100,000. How many lines are in the output?

```{bash, eval=F}
awk '$1 == "chr11"' /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Annotation/Genes/genes.gtf | \ 
  awk '$4 > 100000' | grep -c ""
```

This is the command. I can't get rmarkdown to pass it correctly to bash, and can't figure out the right
syntax to pass it as an r system() function, but the bash code works and returns 54,789 lines.
#MVC#################################################################################
# Odd, that should work. Your answer is right, but just FYI, you can also combine things like this:
```{bash}
  awk '$1 == "chr11" && $4 >= 100000 {print $_}' \
    /reference/homo_sapiens/hg38/ucsc_assembly/illumina_download/Annotation/Genes/genes.gtf \
    | wc -l
```
------------------------------------------------------------------------

# I hate myself - 5 point per question

## Identify the mystery data type

41.  On the Franklin cluster is a folder /gpfs0/scratch/mvc002/. Inside this
folder are two files:
  -   mysteryData_R1.fastq.gz
  -   mysteryData_R2.fastq.gz
  What kind of sequencing experiment are these data derived from?

42.  What species do these reads come from?

No idea, but it's not human!

## Harder programming

43.  Share code that finds a specified sequence in a fasta file, which also
allows for ambiguous bases in the query
(https://droog.gs.washington.edu/parc/images/iupac.html)
